botocore/credentials.py

```diff
@@ -791,6 +820,206 @@ def _assume_role_base_kwargs(self, config):
         return assume_role_kwargs
 
 
+def _role_selector(role_arn, roles):
+    """Given a roles list in the form of [{"RoleArn": "...", ...}, ...],
+    return the item which matches the role_arn, or None otherwise"""
+    chosen = [r for r in roles if r['RoleArn']==role_arn]
+    return chosen[0] if chosen else None
+
+
+class AssumeRoleWithSamlProvider(AssumeRoleProvider):
+    METHOD = 'assume-role-with-saml'
+    DISTINCTION_VAR = 'saml_endpoint'
+
+    def __init__(self, load_config, client_creator, cache, profile_name,
+                 role_selector=_role_selector,
+                 password_prompter=getpass.getpass,
+                 username_prompter=raw_input,
+                 authenticators=None):
+        """To initiate an AssumeRoleWithSamlProvider.
+
+        :type role_selector: callable
+        :param role_selector: A function to choose a role based on both
+            user configuration and the roles allowed by IdP.
+
+        :type password_prompter: callable
+        :param password_prompter: A callable that receives a prompt for user,
+            and then returns a password provided by the user.
+
+        :type username_prompter: callable
+        :param username_prompter: A callable that receives a prompt for user,
+            and then returns the username provided by the user.
+
+        :type authenticators: list
+        :param authenticators: A list of authenticators, which are instances
+            with an is_suitable() method and an authenticate() method.
+            You can use it to add your own implementation for 3rd party IdP.
+        """
+        super(AssumeRoleWithSamlProvider, self).__init__(
+            load_config, client_creator, cache, profile_name)
+        self.username_prompter = username_prompter
+        self.password_prompter = password_prompter
+        self.role_selector = role_selector
+        self.authenticators = authenticators or [
```

Comment:
anytime you have a default value of `None`, you should be using `if authenticators is not None`.  A user may want an empty list of authenticators to start with (maybe they'll inject them later).